pkg/backend/kubernetes/servicemesh/envoy/service_mesh.go

```diff
@@ -340,21 +340,24 @@ func (sm *DefaultEnvoyServiceMesh) ServiceIP(
 	service *latticev1.Service, address *latticev1.Address) (string, map[string]string, error) {
 	ip := address.Annotations[annotationKeyIP]
 
+	protocol := "NULL"
 	protocols := serviceProtocols(service)
-	if len(protocols) != 1 {
-		return "", nil, fmt.Errorf("expected 1 protocol in component ports for service %s, found: %v",
+	if len(protocols) > 1 {
+		return "", nil, fmt.Errorf("expected 0 or 1 protocols in component ports for service %s, found: %v",
 			service.Name, protocols)
+	} else if len(protocols) == 1 {
+		protocol = protocols[0]
 	}
 
-	switch protocols[0] {
+	switch protocol {
 	case "HTTP":
 		netIP := sm.redirectCIDRBlock.IP.String()
 		if ip != "" && ip != netIP {
 			return "", nil, fmt.Errorf("got IP %s for service %s, expected %s", ip, service.Name, netIP)
 		} else {
 			ip = netIP
 		}
-	case "TCP":
+	case "TCP", "NULL":
```

Comment:
what are the consequences of having a service with no ports act like a tcp service?

Reply:
this has only to do with IP assignment. when v2 filter chain matching is implemented, we will have a single listener and this block can be collapsed (i.e., protocol becomes irrelevant here).